packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -707,16 +565,57 @@ contract Bridge is EssentialContract, IBridge {
         }
     }
 
-    function _isPostInvocationDelay(
-        uint256 _receivedAt,
-        uint256 _invocationDelay
+    function _calcFee(
+        uint256 _msgFee,
+        uint256 _msgGasLimit,
+        uint256 _gasUsed
     )
         private
         view
-        returns (bool)
+        returns (uint256)
+    {
+        if (_msgFee == 0 || _msgGasLimit == 0) return 0;
+
+        uint256 maxFee = _msgFee * _gasUsed / _msgGasLimit;
+        uint256 baseFee = block.basefee * _gasUsed;
+        return _msgFee.min(baseFee >= maxFee ? maxFee : (maxFee + baseFee) >> 1);
+    }
+
+    function _calcGasToCharge(
+        uint256 _gasStart,
+        uint256 _gasOverhead,
+        uint256 _calldataBytes,
+        uint256 _numCacheOps
+    )
+        private
+        view
+        returns (uint32)
+    {
+        unchecked {
+            uint256 gasUsed = _gasStart - gasleft() + _gasOverhead
+                + _calldataBytes.min(_MAX_PROOF_BYTES_TO_CHARGE) >> 4;
+
+            uint256 refund = _numCacheOps * _GAS_REFUND_PER_CACHE_OPERATION;
```

Comment:
The user doing the caching will still end up paying the onchain verification cost for the additional merkle proof verification path. Which I guess is okay, the relayer will pick a message that has a max fee that is able to pay this cost for him, but if not the relayer will still be forced to eat the cost of this caching because the max fee for the message will be hit, and then he will have to make up for the cost of this caching he paid himself using the profit of other bridge transactions that are now more efficient/profitable to process (because the reduced cost of the merkle proof verification now allows enough of the fee to be used to actually process the call while still being profitable.

The dynamics here between the two different steps get quite tricky which is why I think it's easier to seem them as completely separate transactions. But the system in this PR where both can still influence each other I think is likely also okay but harder to wrap your head around.

Reply:
<img width="575" alt="Screenshot 2024-04-17 at 11 48 25" src="https://github.com/taikoxyz/taiko-mono/assets/99078276/22ca0249-7ab0-42fe-9d24-99562bb947d1">

Actually caching doesn't trigger additional merkle proof verification path.

Reply:
It doesn't trigger additional merkle proof verification, but the bridge message the relayer uses to do the caching does end up paying to verify the bigger merkle proof. Bridge messages that come afterwards where the relayer makes use of the cached roots then have to verify less merkle proof paths so it will be cheaper for them. So it's not additional, but the the bridge message that did the caching is the sole payer normally of this "normal" bigger merkle proof verification.